telemetry/otlp_exporter.go

```diff
@@ -0,0 +1,166 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"math"
+	"net/http"
+	"time"
+
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	otmetric "go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+)
+
+func StartOtlpExporter(cfg Config) {
+	ctx := context.Background()
+
+	exporter, err := otlpmetricgrpc.New(ctx,
+		otlpmetricgrpc.WithEndpoint(cfg.OtlpCollectorGrpcAddr),
+		otlpmetricgrpc.WithInsecure(),
+	)
+	if err != nil {
+		log.Fatalf("OTLP exporter setup failed: %v", err)
+	}
+
+	res, _ := resource.New(ctx, resource.WithAttributes(
+		semconv.ServiceName(cfg.OtlpServiceName),
+	))
+
+	meterProvider := metric.NewMeterProvider(
+		metric.WithReader(metric.NewPeriodicReader(exporter,
+			metric.WithInterval(15*time.Second))),
+		metric.WithResource(res),
+	)
+	otel.SetMeterProvider(meterProvider)
+	meter := otel.Meter("cosmos-sdk-otlp-exporter")
+
+	gauges := make(map[string]otmetric.Float64Gauge)
+	histograms := make(map[string]otmetric.Float64Histogram)
+
+	go func() {
+		for {
+			if err := scrapeAndPushMetrics(ctx, cfg.PrometheusEndpoint, meter, gauges, histograms); err != nil {
+				log.Printf("error scraping metrics: %v", err)
+			}
+			time.Sleep(15 * time.Second)
+		}
+	}()
```

Comment:
Does the `PeriodicReader` not do this when it calls `run` https://github.com/open-telemetry/opentelemetry-go/blob/9e81492af155454dd6072d6134aa31b91193255b/sdk/metric/periodic_reader.go#L126
?

Reply:
My understanding is the periodicReader does the pushing to the collector, but this function does the reading prometheus metrics and converting them into otlp metrics.

I renamed the function to `scrapePrometheusMetrics` for clarity